### ISSUE TITLE
Heimdal 7.3: KDC bug when cross-realm key is not available

### ISSUE THREAD

**Issue by ahaupt** [heimdal/heimdal#299]
Following up the thread at heimdal-discuss:

http://www.h5l.org/pipermail/heimdal-discuss/2017-July/000190.html

In case a 7.3 KDC does not have a cross-realm key stored, it does not reply with a 'server not found' error but does not respond to the client at all which let's it finally time out.

Setup:

141.34.32.72 -> SL7 client (GSSAPI-enabled system's ssh), located in realm IFH.DE
141.34.22.251 -> Heimdal-7.3 server, only KDC configured on client for IFH.DE

```
[chap-vm1] ~ % ssh -vvv lxplus.cern.ch
OpenSSH_6.6.1, OpenSSL 1.0.1e-fips 11 Feb 2013
[...]
debug1: Next authentication method: gssapi-with-mic
debug1: Unspecified GSS failure.  Minor code may provide more information
Cannot contact any KDC for realm 'IFH.DE'
[...]
Password:
```
This is what is captured as traffic between client and KDC (named test-kdc):

```
[chap-vm1] /root # tshark host test-kdc -t a
Running as user "root" and group "root". This could be dangerous.
Capturing on 'eth0'
  1 16:06:43.529842068 141.34.32.72 -> 141.34.22.251 KRB5 990 TGS-REQ
  2 16:06:44.532691656 141.34.32.72 -> 141.34.22.251 UDP 990 Source port: 48551  Destination port: loadav
  3 16:06:45.533928362 141.34.32.72 -> 141.34.22.251 TCP 74 52164 > kerberos [SYN] Seq=0 Win=29200 Len=0 MSS=1460 SACK_PERM=1 TSval=1035010913 TSecr=0 WS=128
  4 16:06:45.534376809 141.34.22.251 -> 141.34.32.72 TCP 74 kerberos > 52164 [SYN, ACK] Seq=0 Ack=1 Win=28960 Len=0 MSS=1460 SACK_PERM=1 TSval=1816942862 TSecr=1035010913 WS=128
  5 16:06:45.534427289 141.34.32.72 -> 141.34.22.251 TCP 66 52164 > kerberos [ACK] Seq=1 Ack=1 Win=29312 Len=0 TSval=1035010914 TSecr=1816942862
  6 16:06:45.534545716 141.34.32.72 -> 141.34.22.251 KRB5 1018 TGS-REQ
  7 16:06:45.534903298 141.34.22.251 -> 141.34.32.72 TCP 66 kerberos > 52164 [ACK] Seq=1 Ack=953 Win=30976 Len=0 TSval=1816942863 TSecr=1035010914
  8 16:06:45.536931301 141.34.22.251 -> 141.34.32.72 TCP 66 kerberos > 52164 [FIN, ACK] Seq=1 Ack=953 Win=30976 Len=0 TSval=1816942865 TSecr=1035010914
  9 16:06:45.537047128 141.34.32.72 -> 141.34.22.251 TCP 66 52164 > kerberos [FIN, ACK] Seq=953 Ack=2 Win=29312 Len=0 TSval=1035010916 TSecr=1816942865
 10 16:06:45.542447536 141.34.22.251 -> 141.34.32.72 TCP 66 kerberos > 52164 [ACK] Seq=2 Ack=954 Win=30976 Len=0 TSval=1816942870 TSecr=1035010916
 11 16:06:48.536256407 141.34.32.72 -> 141.34.22.251 KRB5 990 TGS-REQ
 12 16:06:49.537370582 141.34.32.72 -> 141.34.22.251 UDP 990 Source port: 48551  Destination port: loadav
 13 16:06:54.542592820 141.34.32.72 -> 141.34.22.251 KRB5 990 TGS-REQ
 14 16:06:55.543675219 141.34.32.72 -> 141.34.22.251 UDP 990 Source port: 48551  Destination port: loadav
```
And the matching KDC logs:

```
Jul 11 16:06:43 chip-vm8 kdc[17992]: TGS-REQ ahaupt@IFH.DE from IPv4:141.34.32.72 for host/lxplus010.cern.ch@IFH.DE [canonicalize, renewable, forwardable]
Jul 11 16:06:43 chip-vm8 kdc[17992]: Searching referral for lxplus010.cern.ch
Jul 11 16:06:43 chip-vm8 kdc[17992]: Server not found in database: krbtgt/CERN.CH@IFH.DE: Success
Jul 11 16:06:44 chip-vm8 kdc[17992]: TGS-REQ ahaupt@IFH.DE from IPv4:141.34.32.72 for host/lxplus010.cern.ch@IFH.DE [canonicalize, renewable, forwardable]
Jul 11 16:06:44 chip-vm8 kdc[17992]: Searching referral for lxplus010.cern.ch
Jul 11 16:06:44 chip-vm8 kdc[17992]: Server not found in database: krbtgt/CERN.CH@IFH.DE: Success
Jul 11 16:06:45 chip-vm8 kdc[17992]: TGS-REQ ahaupt@IFH.DE from IPv4:141.34.32.72 for host/lxplus010.cern.ch@IFH.DE [canonicalize, renewable, forwardable]
Jul 11 16:06:45 chip-vm8 kdc[17992]: Searching referral for lxplus010.cern.ch
Jul 11 16:06:45 chip-vm8 kdc[17992]: Server not found in database: krbtgt/CERN.CH@IFH.DE: Success
Jul 11 16:06:48 chip-vm8 kdc[17992]: TGS-REQ ahaupt@IFH.DE from IPv4:141.34.32.72 for host/lxplus010.cern.ch@IFH.DE [canonicalize, renewable, forwardable]
Jul 11 16:06:48 chip-vm8 kdc[17992]: Searching referral for lxplus010.cern.ch
Jul 11 16:06:48 chip-vm8 kdc[17992]: Server not found in database: krbtgt/CERN.CH@IFH.DE: Success
Jul 11 16:06:49 chip-vm8 kdc[17992]: TGS-REQ ahaupt@IFH.DE from IPv4:141.34.32.72 for host/lxplus010.cern.ch@IFH.DE [canonicalize, renewable, forwardable]
Jul 11 16:06:49 chip-vm8 kdc[17992]: Searching referral for lxplus010.cern.ch
Jul 11 16:06:49 chip-vm8 kdc[17992]: Server not found in database: krbtgt/CERN.CH@IFH.DE: Success
Jul 11 16:06:54 chip-vm8 kdc[17992]: TGS-REQ ahaupt@IFH.DE from IPv4:141.34.32.72 for host/lxplus010.cern.ch@IFH.DE [canonicalize, renewable, forwardable]
Jul 11 16:06:54 chip-vm8 kdc[17992]: Searching referral for lxplus010.cern.ch
Jul 11 16:06:54 chip-vm8 kdc[17992]: Server not found in database: krbtgt/CERN.CH@IFH.DE: Success
Jul 11 16:06:55 chip-vm8 kdc[17992]: TGS-REQ ahaupt@IFH.DE from IPv4:141.34.32.72 for host/lxplus010.cern.ch@IFH.DE [canonicalize, renewable, forwardable]
Jul 11 16:06:55 chip-vm8 kdc[17992]: Searching referral for lxplus010.cern.ch
Jul 11 16:06:55 chip-vm8 kdc[17992]: Server not found in database: krbtgt/CERN.CH@IFH.DE: Success
```

**Comment by vdukhovni** [comment]
IIRC this was fixed.  Please re-open if still applicable.

**Comment by nicowilliams** [comment]
@madscientist159 thanks for the report.  The fix will be in Heimdal 8.0, and is in master at this time.
